File: go-backend/download.go
package main

type Download struct {
	AppId        string  `json:"app_id"`
	Latitude     float64 `json:"latitude"`
	Longitude    float64 `json:"longitude"`
	DownloadedAt string  `json:"downloaded_at"`
	Country      string  `json:"country"`
}

func filterByDate(ss []Download, from string, to string, test func(string, string, string) bool) (ret []Download) {
	for _, s := range ss {
		if test(s.DownloadedAt, from, to) {
			ret = append(ret, s)
		}
	}
	return
}

File: app/src/services/download.ts
export async function addRandomDownload() {
  const headers = new Headers();
  headers.append("Content-Type", "application/json");

  const response = await fetch(`http://${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}/download`, {
    headers,
    method: "POST",
    mode: "cors",
    cache: "default",
    body: JSON.stringify({
      app_id: Math.random() > 0.5 ? "IOS" : "Android",
      downloaded_at: new Date().toUTCString(),
      latitude: (-100.1 + Math.random() * 9000.0) / 100,
      longitude: (100.1 + Math.random() * 9000.0) / 100
    })
  });

  return response;
}

export async function getAllDownloads(filter?: { from: Date, to: Date }) {
  const headers = new Headers();
  headers.append("Content-Type", "application/json");
  const queryString = filter ? `?from=${filter.from.toUTCString()}&to=${filter.to.toUTCString()}` : "";
  const response = await fetch(`http://${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}/downloads${queryString}`, {
    headers,
    method: "GET",
    mode: "cors",
    cache: "default"
  });
  return response;
}